File: core/src/main/scala/eu/ostrzyciel/jelly/core/internal/ProtoEncoderImpl.scala
package eu.ostrzyciel.jelly.core.internal

import eu.ostrzyciel.jelly.core.*
import eu.ostrzyciel.jelly.core.proto.v1.*

import scala.collection.mutable

private[core] object ProtoEncoderImpl:
  private val graphEnd = RdfStreamRow(RdfGraphEnd.defaultInstance)
  private val defaultGraphStart = RdfStreamRow(RdfGraphStart(RdfDefaultGraph.defaultInstance))
  private val emptyRowBuffer: List[RdfStreamRow] = List()

/**
 * Stateful encoder of a protobuf RDF stream.
 *
 * This class supports all stream types and options, but usually does not check if the user is conforming to them.
 * It will, for example, allow the user to send generalized triples in a stream that should not have them.
 * Take care to ensure the correctness of the transmitted data, or use the specialized wrappers from the stream package.
 *
 * @param converter converter for the encoder
 * @param params parameters object for the encoder
 */
private[core] final class ProtoEncoderImpl[TNode, -TTriple, -TQuad](
  protected val converter: ProtoEncoderConverter[TNode, TTriple, TQuad],
  params: ProtoEncoder.Params,
) extends ProtoEncoder[TNode, TTriple, TQuad, ?]:

  import ProtoEncoderImpl.*
  
  override val enableNamespaceDeclarations: Boolean = params.enableNamespaceDeclarations
  // Override whatever the user set in the options.
  override val options: RdfStreamOptions = params.options.withVersion(
    // If namespace declarations are enabled, we need to use Jelly 1.1.x.
    if enableNamespaceDeclarations then Constants.protoVersion_1_1_x
    else Constants.protoVersion_1_0_x
  )
  override val maybeRowBuffer: Option[mutable.Buffer[RdfStreamRow]] = params.maybeRowBuffer

  /** @inheritdoc */
  override def addTripleStatement(
    subject: TNode, predicate: TNode, `object`: TNode
  ): Iterable[RdfStreamRow] =
    handleHeader()
    val mainRow = RdfStreamRow(tripleToProto(subject, predicate, `object`))
    appendAndReturn(mainRow)

  /** @inheritdoc */
  override def addQuadStatement(
    subject: TNode, predicate: TNode, `object`: TNode, graph: TNode
  ): Iterable[RdfStreamRow] =
    handleHeader()
    val mainRow = RdfStreamRow(quadToProto(subject, predicate, `object`, graph))
    appendAndReturn(mainRow)

  /** @inheritdoc */
  override def startGraph(graph: TNode): Iterable[RdfStreamRow] =
    if graph == null then
      startDefaultGraph()
    else
      handleHeader()
      val graphNode = converter.graphNodeToProto(nodeEncoder, graph)
      val mainRow = RdfStreamRow(RdfGraphStart(graphNode))
      appendAndReturn(mainRow)

  /** @inheritdoc */
  override def startDefaultGraph(): Iterable[RdfStreamRow] =
    handleHeader()
    appendAndReturn(defaultGraphStart)

  /** @inheritdoc */
  override def endGraph(): Iterable[RdfStreamRow] =
    if !emittedOptions then
      throw new RdfProtoSerializationError("Cannot end a delimited graph before starting one")
    appendAndReturn(graphEnd)

  /** @inheritdoc */
  override def declareNamespace(name: String, iriValue: String): Iterable[RdfStreamRow] =
    if !enableNamespaceDeclarations then
      throw new RdfProtoSerializationError("Namespace declarations are not enabled in this stream")
    handleHeader()
    val mainRow = RdfStreamRow(RdfNamespaceDeclaration(
      name,
      nodeEncoder.makeIri(iriValue).iri
    ))
    appendAndReturn(mainRow)

  // *** PRIVATE FIELDS AND METHODS ***
  // **********************************
  private val rowBuffer: mutable.Buffer[RdfStreamRow] = maybeRowBuffer.getOrElse(mutable.ListBuffer[RdfStreamRow]())
  // Whether the encoder is responsible for clearing the buffer.
  private val iResponsibleForBufferClear: Boolean = maybeRowBuffer.isEmpty
  override protected val nodeEncoder: NodeEncoderImpl[TNode] = new NodeEncoderImpl[TNode](
    options,
    this, // RowBufferAppender
    // Make the node cache size between 256 and 1024, depending on the user's maxNameTableSize.
    Math.max(Math.min(options.maxNameTableSize, 1024), 256),
    options.maxNameTableSize,
    Math.max(Math.min(options.maxNameTableSize, 1024), 256),
  )
  private var emittedOptions: Boolean = false

  private[core] override def appendLookupEntry(entry: RdfLookupEntryRowValue): Unit =
    rowBuffer.append(RdfStreamRow(entry))

  private inline def handleHeader(): Unit =
    if !emittedOptions then emitOptions()

  private def appendAndReturn(row: RdfStreamRow): Iterable[RdfStreamRow] =
    rowBuffer.append(row)
    // This branch will always be correctly predicted
    if iResponsibleForBufferClear then
      val list = rowBuffer.toList
      rowBuffer.clear()
      list
    else emptyRowBuffer

  private def emitOptions(): Unit =
    emittedOptions = true
    rowBuffer.append(RdfStreamRow(options))

File: core/src/main/java/eu/neverblink/jelly/core/internal/ProtoEncoderImpl.java
package eu.neverblink.jelly.core.internal;

import eu.neverblink.jelly.core.*;
import eu.neverblink.jelly.core.ProtoEncoder;
import eu.neverblink.jelly.core.ProtoEncoderConverter;
import eu.neverblink.jelly.core.RdfProtoSerializationError;
import eu.neverblink.jelly.core.proto.v1.*;

/**
 * Stateful encoder of a protobuf RDF stream.
 * <p>
 * This class supports all stream types and options, but usually does not check if the user is conforming to them.
 * It will, for example, allow the user to send generalized triples in a stream that should not have them.
 * Take care to ensure the correctness of the transmitted data, or use the specialized wrappers from the stream package.
 */
@InternalApi
public class ProtoEncoderImpl<TNode> extends ProtoEncoder<TNode> {

    private boolean hasEmittedOptions = false;

    /**
     * Constructor for the ProtoEncoderImpl class.
     * <p>
     *
     * @param converter converter for the encoder
     * @param params    parameters object for the encoder
     */
    public ProtoEncoderImpl(ProtoEncoderConverter<TNode> converter, ProtoEncoder.Params params) {
        super(converter, params);
    }

    @Override
    public void handleTriple(TNode subject, TNode predicate, TNode object) {
        emitOptions();
        final var triple = tripleToProto(subject, predicate, object);
        // Calculate the size of the row now, as all objects are likely still in L1/L2 cache.
        rowBuffer.appendMessage().setTriple(triple).getSerializedSize();
    }

    @Override
    public void handleQuad(TNode subject, TNode predicate, TNode object, TNode graph) {
        emitOptions();
        final var quad = quadToProto(subject, predicate, object, graph);
        // Calculate the size of the row now, as all objects are likely still in L1/L2 cache.
        rowBuffer.appendMessage().setQuad(quad).getSerializedSize();
    }

    @Override
    public void handleGraphStart(TNode graph) {
        emitOptions();
        final var graphStart = graphStartToProto(graph);
        rowBuffer.appendMessage().setGraphStart(graphStart).getSerializedSize();
    }

    @Override
    public void handleGraphEnd() {
        if (!hasEmittedOptions) {
            throw new RdfProtoSerializationError("Cannot end a delimited graph before starting one");
        }
        rowBuffer.appendMessage().setGraphEnd(RdfGraphEnd.EMPTY).getSerializedSize();
    }

    @Override
    public void handleNamespace(String prefix, TNode namespace) {
        if (!enableNamespaceDeclarations) {
            throw new RdfProtoSerializationError("Namespace declarations are not enabled in this stream");
        }

        emitOptions();

        final var ns = RdfNamespaceDeclaration.newInstance().setName(prefix);
        final var encoded = converter.nodeToProto(getNodeEncoder(), namespace);
        ns.setValue((RdfIri) encoded.node());
        rowBuffer.appendMessage().setNamespace(ns).getSerializedSize();
    }

    @Override
    public void appendNameEntry(RdfNameEntry nameEntry) {
        rowBuffer.appendMessage().setName(nameEntry).getSerializedSize();
    }

    @Override
    public void appendPrefixEntry(RdfPrefixEntry prefixEntry) {
        rowBuffer.appendMessage().setPrefix(prefixEntry).getSerializedSize();
    }

    @Override
    public void appendDatatypeEntry(RdfDatatypeEntry datatypeEntry) {
        rowBuffer.appendMessage().setDatatype(datatypeEntry).getSerializedSize();
    }

    private void emitOptions() {
        if (hasEmittedOptions) {
            return;
        }

        hasEmittedOptions = true;
        rowBuffer.appendMessage().setOptions(options).getSerializedSize();
    }
}
